pkg/services/thumbs/crawler_auth.go

```diff
@@ -2,40 +2,137 @@ package thumbs
+	sqlStore                 *sqlstore.SQLStore
+}
 
 type CrawlerAuth interface {
 	GetUserId(orgId int64) int64
+	GetLogin(orgId int64) string
 	GetOrgRole() models.RoleType
 }
 
-type staticCrawlerAuth struct {
-	userId  int64
-	orgRole models.RoleType
+func (o *OSSCrawlerAuthSetupService) findAllOrgIds(ctx context.Context) ([]int64, error) {
+	searchAllOrgsQuery := &models.SearchOrgsQuery{}
+	if err := o.sqlStore.SearchOrgs(ctx, searchAllOrgsQuery); err != nil {
+		o.log.Error("Error when searching for orgs", "err", err)
+		return nil, err
+	}
+
+	orgIds := make([]int64, 0)
+	for i := range searchAllOrgsQuery.Result {
+		orgIds = append(orgIds, searchAllOrgsQuery.Result[i].Id)
+	}
+
+	return orgIds, nil
+}
+
+type crawlerAuth struct {
+	accountIdByOrgId map[int64]int64
+	loginByOrgId     map[int64]string
+	orgRole          models.RoleType
 }
 
-func (o *staticCrawlerAuth) GetOrgRole() models.RoleType {
+func (o *crawlerAuth) GetOrgRole() models.RoleType {
 	return o.orgRole
 }
 
-func (o *staticCrawlerAuth) GetUserId(orgId int64) int64 {
-	return o.userId
+func (o *crawlerAuth) GetUserId(orgId int64) int64 {
+	return o.accountIdByOrgId[orgId]
+}
+
+func (o *crawlerAuth) GetLogin(orgId int64) string {
+	return o.loginByOrgId[orgId]
 }
 
 func (o *OSSCrawlerAuthSetupService) Setup(ctx context.Context) (CrawlerAuth, error) {
+	orgIds, err := o.findAllOrgIds(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	// userId:0 and ROLE_ADMIN grants the crawler process permissions to view all dashboards in all folders & orgs
 	// the process doesn't and shouldn't actually need to edit/modify any resources from the UI
-	return &staticCrawlerAuth{userId: 0, orgRole: models.ROLE_ADMIN}, nil
+	orgRole := models.ROLE_ADMIN
+
+	accountIdByOrgId := make(map[int64]int64)
+	loginByOrgId := make(map[int64]string)
+	for _, orgId := range orgIds {
+		o.log.Info("Creating account for org", "orgId", orgId)
+
+		serviceAccountNameOrg := o.serviceAccountNamePrefix + strconv.FormatInt(orgId, 10)
+
+		saForm := serviceaccounts.CreateServiceAccountForm{
+			Name: serviceAccountNameOrg,
+			Role: &orgRole,
+		}
+
+		serviceAccount, err := o.serviceAccounts.CreateServiceAccount(context.Background(), orgId, &saForm)
+		accountAlreadyExists := errors.Is(err, database.ErrServiceAccountAlreadyExists)
+
+		if !accountAlreadyExists && err != nil {
+			o.log.Error("Failed to create the service account", "err", err, "accountName", serviceAccountNameOrg, "orgId", orgId)
+			return nil, err
+		}
+
+		var serviceAccountLogin string
+		var serviceAccountId int64
+		if accountAlreadyExists {
+			id, err := o.serviceAccounts.RetrieveServiceAccountIdByName(ctx, orgId, serviceAccountNameOrg)
+			if err != nil {
+				o.log.Error("Failed to retrieve service account", "err", err, "accountName", serviceAccountNameOrg)
+				return nil, err
+			}
+
+			// update org_role to make sure everything works properly if someone has changed the role since SA's original creation
```

Comment:
Hm, maybe this is too smart, if the role has been downgraded explicitly then there could be a reason for it?

Reply:
Well, it won't work otherwise.. :) This is a safeguard against accidental change & a piece of code that allows for smooth transition between licensed, unlicensed and OSS versions. I'll add a config option to disable the crawler altogether if someone doesn't want it to run.